modules/pam_issue/pam_issue.8
.\"     Title: pam_issue
.\"    Author: 
.\" Generator: DocBook XSL Stylesheets v1.70.1 <http://docbook.sf.net/>
.\"      Date: 06/17/2006
.\"    Manual: Linux\-PAM Manual
.\"    Source: Linux\-PAM Manual
.\"
.TH "PAM_ISSUE" "8" "06/17/2006" "Linux\-PAM Manual" "Linux\-PAM Manual"
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.SH "NAME"
pam_issue \- PAM module to add issue file to user prompt
.SH "SYNOPSIS"
.HP 13
\fBpam_issue.so\fR [noesc] [issue=\fIissue\-file\-name\fR]
.SH "DESCRIPTION"
.PP
pam_issue is a PAM module to prepend an issue file to the username prompt. It also by default parses escape codes in the issue file similar to some common getty's (using \\x format).
.PP
Recognized escapes:
.TP 3n
\fB\\d\fR
current day
.TP 3n
\fB\\l\fR
name of this tty
.TP 3n
\fB\\m\fR
machine architecture (uname \-m)
.TP 3n
\fB\\n\fR
machine's network node hostname (uname \-n)
.TP 3n
\fB\\o\fR
domain name of this system
.TP 3n
\fB\\r\fR
release number of operating system (uname \-r)
.TP 3n
\fB\\t\fR
current time
.TP 3n
\fB\\s\fR
operating system name (uname \-s)
.TP 3n
\fB\\u\fR
number of users currently logged in
.TP 3n
\fB\\U\fR
same as \\u except it is suffixed with "user" or "users" (eg. "1 user" or "10 users")
.TP 3n
\fB\\v\fR
operating system version and build date (uname \-v)
.SH "OPTIONS"
.PP
.TP 3n
\fBnoesc\fR
Turns off escape code parsing.
.TP 3n
\fBissue=\fR\fB\fIissue\-file\-name\fR\fR
The file to output if not using the default.
.SH "MODULE SERVICES PROVIDED"
.PP
Only the
\fBauth\fR
service is supported.
.SH "RETURN VALUES"
.PP
.TP 3n
PAM_BUF_ERR
Memory buffer error.
.TP 3n
PAM_IGNORE
The prompt was already changed.
.TP 3n
PAM_SERVICE_ERR
A service module error occured.
.TP 3n
PAM_SUCCESS
The new prompt was set successfull.
.SH "EXAMPLES"
.PP
Add the following line to
\fI/etc/pam.d/login\fR
to set the user specific issue at login:
.sp
.RS 3n
.nf
        auth optional pam_issue.so issue=/etc/issue
      
.fi
.RE
.sp
.SH "SEE ALSO"
.PP

\fBpam.conf\fR(5),
\fBpam.d\fR(8),
\fBpam\fR(8)
.SH "AUTHOR"
.PP
pam_issue was written by Ben Collins <bcollins@debian.org>.


NEWS
Linux-PAM NEWS -- history of user-visible changes.


* pam_tally: Fix support for large UIDs
* Fixed all problems found by Coverity
* Add support for Intel C Compiler
* Add manual page for pam_mkhomedir, pam_umask, pam_filter,
  pam_issue, pam_ftp, pam_group, pam_lastlog, pam_listfile,
  pam_localuser, pam_mail, pam_motd, pam_nologin, pam_permit,
  pam_rootok, pam_securetty, pam_shells, pam_userdb, pam_warn,
  pam_time, pam_limits, pam_debug, pam_tally
* The libpam memory debug code was removed

Release 0.99.4.0

* Add test suite
* Fix building of static variants of libpam, libpamc and libpam_misc
* pam_listfile: Add support for password and session management
* pam_exec: New PAM module to execute arbitary commands
* Fix building of a static libpam including all PAM modules
* New/updated translations for: nl, pt, pl, fi, km, tr, uk, fr
* pam_access: Add network(address) / netmask and IPv6 support
* Add manual pages for pam_cracklib, pam_deny and pam_access
* pam_pwdb: This deprecated module was removed
* Manual pages: Major rewrite/cleanup


Release 0.99.3.0

* Fix NULL pointer checks in libpam.so
* pam_succeed_if, pam_group, pam_time: Support netgroup matching
* New translations for: nb, hu, fi, de, es, fr, it, ja, pt_BR, zh_CN, zh_TW
* Audit PAM calls if Linux Audit is available
* Compile upperLOWER and unix_chkpwd as PIE binaries


Release 0.99.2.1

* Fix install of PS, PDF, TXT and HTML files
* pam_mail: Update README
* Use %m consistent
* pam_modutil_getlogin: Fix parsing of PAM_TTY variable


Release 0.99.2.0

* Fix parsing of full path tty name in various modules
* pam_xauth: Look for xauth executable in multiple places
* pam_unix: Disable user check in unix_chkpwd only if real uid
  is 0 (CVE-2005-2977). Log failed password check attempt.
* pam_env: Support /etc/environment again, but don't treat it as
  error if it is missing.
* pam_userdb: Fix memory leak.


Release 0.99.1.0

* Use autoconf/automake/libtool
* Add gettext support
* Add translations for cs, de, es, fr, hu, it, ja, nb, pa, pt_BR,
  pt, zh_CN and zh_TW
* libpam: Remove pam_authenticate_secondary stub
* libpam: Add pam_prompt,pam_vprompt,pam_error,pam_verror,pam_info
  and pam_vinfo functions for use by modules as extension
* libpam: Add pam_syslog function for unified syslog messages from
  PAM modules
* libpam: Moved functions from pammodutil to libpam
* pam_umask: New module for setting umask from GECOS field, /etc/login.defs
  or /etc/default/login
* pam_echo: New PAM module for message output
* pam_userdb: Fix regression (crash when crypt param not specified)
* pam_limits: Fix regression from RLIMIT_NICE support (wrong limit
  values for other limits are applied)
* pam_access: Support for NULL tty - matches ALL and NONE keywords
* pam_lastlog: Enable log to wtmp by default.  Add "nowtmp" option
* pam_radius: This module was removed


modules/pam_issue/pam_issue.c
/* pam_issue module - a simple /etc/issue parser to set PAM_USER_PROMPT
 *
 * Copyright 1999 by Ben Collins <bcollins@debian.org>
 *
 * Needs to be called before any other auth modules so we can setup the
 * user prompt before it's first used. Allows one argument option, which
 * is the full path to a file to be used for issue (uses /etc/issue as a
 * default) such as "issue=/etc/issue.telnet".
 *
 * We can also parse escapes within the the issue file (enabled by
 * default, but can be disabled with the "noesc" option). It's the exact
 * same parsing as util-linux's agetty program performs.
 *
 * Released under the GNU LGPL version 2 or later
 */

#include "config.h"

#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <string.h>
#include <unistd.h>
#include <sys/utsname.h>
#include <utmp.h>
#include <time.h>
#include <syslog.h>

#define PAM_SM_AUTH

#include <security/_pam_macros.h>
#include <security/pam_modules.h>
#include <security/pam_ext.h>

static int _user_prompt_set = 0;

static int read_issue_raw(pam_handle_t *pamh, FILE *fp, char **prompt);
static int read_issue_quoted(pam_handle_t *pamh, FILE *fp, char **prompt);

/* --- authentication management functions (only) --- */

PAM_EXTERN int
pam_sm_authenticate (pam_handle_t *pamh, int flags UNUSED,
		     int argc, const char **argv)
{
    int retval = PAM_SERVICE_ERR;
    FILE *fp;
    const char *issue_file = NULL;
    int parse_esc = 1;
    const void *item = NULL;
    const char *cur_prompt;
    char *issue_prompt = NULL;

   /* If we've already set the prompt, don't set it again */
    if(_user_prompt_set)
	return PAM_IGNORE;

    /* We set this here so if we fail below, we wont get further
       than this next time around (only one real failure) */
    _user_prompt_set = 1;

    for ( ; argc-- > 0 ; ++argv ) {
	if (!strncmp(*argv,"issue=",6)) {
	    issue_file = 6 + *argv;
	    D(("set issue_file to: %s", issue_file));
	} else if (!strcmp(*argv,"noesc")) {
	    parse_esc = 0;
	    D(("turning off escape parsing by request"));
	} else
	    D(("unknown option passed: %s", *argv));
    }

    if (issue_file == NULL)
	issue_file = "/etc/issue";

    if ((fp = fopen(issue_file, "r")) == NULL) {
	pam_syslog(pamh, LOG_ERR, "error opening %s: %m", issue_file);
	return PAM_SERVICE_ERR;
    }

    if ((retval = pam_get_item(pamh, PAM_USER_PROMPT, &item)) != PAM_SUCCESS) {
	fclose(fp);
	return retval;
    }

    cur_prompt = item;
    if (cur_prompt == NULL)
	cur_prompt = "";

    if (parse_esc)
	retval = read_issue_quoted(pamh, fp, &issue_prompt);
    else
	retval = read_issue_raw(pamh, fp, &issue_prompt);

    fclose(fp);

    if (retval != PAM_SUCCESS)
	goto out;

    {
	size_t size = strlen(issue_prompt) + strlen(cur_prompt) + 1;
	char *new_prompt = realloc(issue_prompt, size);

	if (new_prompt == NULL) {
	    pam_syslog(pamh, LOG_ERR, "out of memory");
	    retval = PAM_BUF_ERR;
	    goto out;
	}
	issue_prompt = new_prompt;
    }

    strcat(issue_prompt, cur_prompt);
    retval = pam_set_item(pamh, PAM_USER_PROMPT,
			      (const void *) issue_prompt);
  out:
    _pam_drop(issue_prompt);
    return (retval == PAM_SUCCESS) ? PAM_IGNORE : retval;
}

PAM_EXTERN int
pam_sm_setcred (pam_handle_t *pamh UNUSED, int flags UNUSED,
		int argc UNUSED, const char **argv UNUSED)
{
     return PAM_IGNORE;
}

static int
read_issue_raw(pam_handle_t *pamh, FILE *fp, char **prompt)
{
    char *issue;
    struct stat st;

    *prompt = NULL;

    if (fstat(fileno(fp), &st) < 0) {
	pam_syslog(pamh, LOG_ERR, "stat error: %m");
	return PAM_SERVICE_ERR;
    }

    if ((issue = malloc(st.st_size + 1)) == NULL) {
	pam_syslog(pamh, LOG_ERR, "out of memory");
	return PAM_BUF_ERR;
    }

    if (fread(issue, 1, st.st_size, fp) != st.st_size) {
	pam_syslog(pamh, LOG_ERR, "read error: %m");
	_pam_drop(issue);
	return PAM_SERVICE_ERR;
    }

    issue[st.st_size] = '\0';
    *prompt = issue;
    return PAM_SUCCESS;
}

static int
read_issue_quoted(pam_handle_t *pamh, FILE *fp, char **prompt)
{
    int c;
    size_t size = 1024;
    char *issue;
    struct utsname uts;

    *prompt = NULL;

    if ((issue = malloc(size)) == NULL) {
	pam_syslog(pamh, LOG_ERR, "out of memory");
	return PAM_BUF_ERR;
    }

    issue[0] = '\0';
    (void) uname(&uts);

    while ((c = getc(fp)) != EOF) {
	char buf[1024];

	buf[0] = '\0';
	if (c == '\\') {
	    if ((c = getc(fp)) == EOF)
		break;
	    switch (c) {
	      case 's':
		strncat(buf, uts.sysname, sizeof(buf) - 1);
		break;
	      case 'n':
		strncat(buf, uts.nodename, sizeof(buf) - 1);
		break;
	      case 'r':
		strncat(buf, uts.release, sizeof(buf) - 1);
		break;
	      case 'v':
		strncat(buf, uts.version, sizeof(buf) - 1);
		break;
	      case 'm':
		strncat(buf, uts.machine, sizeof(buf) - 1);
		break;
	      case 'o':
		{
		    char domainname[256];

		    if (getdomainname(domainname, sizeof(domainname)) >= 0) {
			domainname[sizeof(domainname)-1] = '\0';
			strncat(buf, domainname, sizeof(buf) - 1);
		    }
		}
		break;
	      case 'd':
	      case 't':
		{
		    const char *weekday[] = {
			"Sun", "Mon", "Tue", "Wed", "Thu",
			"Fri", "Sat" };
		    const char *month[] = {
			"Jan", "Feb", "Mar", "Apr", "May",
			"Jun", "Jul", "Aug", "Sep", "Oct",
			"Nov", "Dec" };
		    time_t now;
		    struct tm *tm;

		    (void) time (&now);
		    tm = localtime(&now);

		    if (c == 'd')
			snprintf (buf, sizeof buf, "%s %s %d  %d",
				weekday[tm->tm_wday], month[tm->tm_mon],
				tm->tm_mday, tm->tm_year + 1900);
		    else
			snprintf (buf, sizeof buf, "%02d:%02d:%02d",
				tm->tm_hour, tm->tm_min, tm->tm_sec);
		}
		break;
	      case 'l':
		{
		    char *ttyn = ttyname(1);
		    if (ttyn) {
			if (!strncmp(ttyn, "/dev/", 5))
			    ttyn += 5;
			strncat(buf, ttyn, sizeof(buf) - 1);
		    }
		}
		break;
	      case 'u':
	      case 'U':
		{
		    unsigned int users = 0;
		    struct utmp *ut;
		    setutent();
		    while ((ut = getutent())) {
			if (ut->ut_type == USER_PROCESS)
			    ++users;
		    }
		    endutent();
		    if (c == 'U')
			snprintf (buf, sizeof buf, "%u %s", users,
			          (users == 1) ? "user" : "users");
		    else
			snprintf (buf, sizeof buf, "%u", users);
		    break;
		}
	      default:
		buf[0] = c; buf[1] = '\0';
	    }
	} else {
	    buf[0] = c; buf[1] = '\0';
	}

	if ((strlen(issue) + strlen(buf)) + 1 > size) {
	    char *new_issue;

	    size += strlen(buf) + 1;
	    new_issue = (char *) realloc (issue, size);
	    if (new_issue == NULL) {
		_pam_drop(issue);
		return PAM_BUF_ERR;
	    }
	    issue = new_issue;
	    strcat(issue, buf);
	}
    }

    if (ferror(fp)) {
	pam_syslog(pamh, LOG_ERR, "read error: %m");
	_pam_drop(issue);
	return PAM_SERVICE_ERR;
    }

    *prompt = issue;
    return PAM_SUCCESS;
}

#ifdef PAM_STATIC

/* static module data */

struct pam_module _pam_issue_modstruct = {
     "pam_issue",
     pam_sm_authenticate,
     pam_sm_setcred,
     NULL,
     NULL,
     NULL,
     NULL,
};

#endif

/* end of module definition */
